### DISCUSSION TITLE
How does caching work?

### DISCUSSION THREAD

**Discussion by szmarci** [moonrepo/moon#1470] [Q&A]
I am tinkering with the example repo: [https://github.com/moonrepo/examples](https://github.com/moonrepo/examples)

When I first run `moon run :build`, I get `Tasks: 12 completed`, which is expected it is the first run.

Then I go and change something in the `browser-package`:

```typescript
export function add(a: number, b: number): number {
  return a + b + 2;
}
```

I added the `+ 2`.

I run `moon run :build` again, I get `12 completed (6 cached)`, which is also fine, cause 5 project references this package + the package itself changed, that accounts for the 6 rebuild.

Then I revert my changes in the `browser-package` and run `moon run :build`.
This time I get `12 completed (11 cached)` and according to the logs, the nextjs app was the one that was rebuilt.

`▪▪▪▪ nextjs-app:build (8s 156ms, 9a8edaaf)`

Shouldn't all come from the cache? The code is in the same state it was before the change I did.
Why nextjs is rebuilt?

**Comment by milesj** [comment]
@szmarci You would need to compare the hashes of both next builds to determine why. It's most likely that 1 of the inputs slightly changed.

You can compare it with `moon query hash-diff <a hash> <b hash>`.

**Reply by szmarci** [reply]
browser-package hash after first run:
`69c08924`

browser-package hash after modifying browser-package:
`b53fbf97`

Running `moon query hash-diff 69c08924 b53fbf97`:

```diff
{
  "inputs": {
        "packages/browser-package/src/index.ts": "1a2e299b8cd8e14a682686761bf3e80183154c48",
        "packages/browser-package/src/index.ts": "93b95ee7262a6fc74bb3f3715938ab7db41d7fb9"
  }
}
```

Makes sense, the file changed, I added `return a + b + 1`.

nextjs hash after first run:
`9a8edaaf`

nextjs hash after modifying browser-package:
`ef5f4c9a`

Running `moon query hash-diff 9a8edaaf ef5f4c9a`

```
"deps": {
  "browser-package:build": "69c08924a79d42c7d5839ccb7b3bb2a11d71f111a451b02dac9b7269a5806c03"
  "browser-package:build": "b53fbf97de327db46b68bd7f6a85b1a65df8f78a6ed5e1ca5cfce3d53dcff297"
}
```

Still good, the dependency changed.

Then I revert the browser package to what it was `return a + b`.

In the logs I see:
`▪▪▪▪ nextjs-app:build (8s 815ms, 9a8edaaf)`

It has the same hash after the first run, but it is built again, not like the other packages that has cached in the logs (`▪▪▪▪ astro-app:build (cached, b2647493)`)

In theory when I first run the build the repo is in state A. Then I change something in the browser-packages it is in state B.
And lastly I revert my changes, the repo goes back to state A, everything should come from the cache, right?

If I remove nextjs from the monorepo, it works as expected. After reverting the changes in the `browser-package` everything comes from the cache. 

Is this some nextjs sideeffect?

**Reply by milesj** [reply]
Hrmm nice find. If the hashes are the same, then this may actually be a bug. Kind of seems like a race condition.

I'm actually working on a new task runner so I'll be sure to verify this. https://github.com/moonrepo/moon/pull/1463
